### ISSUE TITLE
[helm 3] How to get Capabilities.KubeVersion.Platform ?

### ISSUE THREAD

**Issue by mmouly** [helm/helm#7004]
Hello,

We currently use Capabilities.KubeVersion.Platform to get OS info.
Now, helm 3 is generating the following error :
executing "arch" at <.Capabilities.KubeVersion.Platform>: can't evaluate field Platform in type chartutil.KubeVersion

Does it exist a replacement ?
Thanks for your help

**Comment by bacongobbler** [comment]
Looks like a few fields were removed that could be added back: https://github.com/helm/helm/blob/0ca95341c19fcaf9cecdf50caef1693568835a84/pkg/chartutil/capabilities.go#L37-L39

**Comment by hickeyma** [comment]
Looks like the docs might need updating as well: https://helm.sh/docs/topics/chart_template_guide/builtin_objects/

**Comment by karuppiah7890** [comment]
Oh, yes. I'll create an issue in docs repo? And link to the PR and issue? So that once the PR is merged, we can update the docs

**Comment by hickeyma** [comment]
Thanks @karuppiah7890. Probably check if the current doc is correct also. It maybe slightly inaccurate.

**Comment by karuppiah7890** [comment]
Sure! I will check it out sometime :)

**Comment by karuppiah7890** [comment]
@hickeyma , @PYLochou pointed out a problem in `Capabilities` doc and I checked it. Raised issue https://github.com/helm/helm-www/issue/414 and PR https://github.com/helm/helm-www/pull/415 for it. And I think docs PR for #7004 issue can be raised after #7062 is merged

**Comment by technosophos** [comment]
In Helm 3, GoVersion and Compiler were intentionally removed as (a) we found no evidence that they were ever used, and (b) they didn't really provide any helpful functionality. (The OS/Arch of the Helm client weren't important, though there was arguably a reason they were important with Tiller).

Platform was probably removed for the same reason (since it was the platform the compiler targeted), without anyone realizing that these might also be used to tell what the host OS/Arch was.

**Comment by adamreese** [comment]
@mmouly What are you using `Platform` for in your chart?

**Comment by mmouly** [comment]
for nodeAffinity , to select the relevant images to a dedicated platform


apiVersion: apps/v1
  kind: Deployment

affinity:
--
  | nodeAffinity:
  | requiredDuringSchedulingIgnoredDuringExecution:
  | #If you specify multiple nodeSelectorTerms associated with nodeAffinity types,
  | #then the pod can be scheduled onto a node if one of the nodeSelectorTerms is satisfied.
  | #
  | #If you specify multiple matchExpressions associated with nodeSelectorTerms,
  | #then the pod can be scheduled onto a node only if all matchExpressions can be satisfied.
  | #
  | #valid operators: In, NotIn, Exists, DoesNotExist, Gt, Lt
  | nodeSelectorTerms:
  | - matchExpressions:
  | - key: beta.kubernetes.io/arch
  | operator: In
  | values:
  | {{- if .Values.image.arch }}
  | - {{ .Values.image.arch }}
  | {{- else }}
  | - {{ template "arch" . }}
  | {{- end }}
  | {{- if not (empty (.Values.customization.dedicatedNodeLabel))  }}
  | - key: {{ .Values.customization.dedicatedNodeLabel }}
  | operator: "Exists"
  | {{- end }}


{{/*
--
  | Return arch based on kube platform
  | */}}
  | {{- define "arch" -}}
  | {{- if (eq "linux/amd64" .Capabilities.KubeVersion.Platform) }}
  | {{- printf "%s" "amd64" }}
  | {{- else if (eq "linux/ppc64le" .Capabilities.KubeVersion.Platform) }}
  | {{- printf "%s" "ppc64le" }}
  | {{- else if (eq "linux/s390x" .Capabilities.KubeVersion.Platform) }}
  | {{- printf "%s" "s390x" }}
  | {{- else }}
  | {{- printf "%s" "amd64" }}
  | {{- end -}}
  | {{- end -}}

**Comment by PriyaVasudevan** [comment]
Is there still a problem with .Capabilities.KubeVersion.Platform ? I am using Helm 3.2.0 and am facing this issue when I include nodeaffinity

Error: template: ibm-db2u/templates/db2u-statefulset.yaml:71:10: executing "ibm-db2u/templates/db2u-statefulset.yaml" at <include "nodeaffinity" .>: error calling include: template: ibm-db2u/templates/_affinity.tpl:7:8: executing "nodeaffinity" at <include "nodeAffinityRequiredDuringScheduling" .>: error calling include: template: ibm-db2u/templates/_architecture.tpl:13:40: executing "helperplatform" at <.Capabilities.KubeVersion.Platform>: can't evaluate field Platform in type chartutil.KubeVersion

**Comment by mmouly** [comment]
yes. I assume that somebody should solve it for V3.3.0 : 
https://github.com/helm/helm/milestone/89

No ?

**Comment by bacongobbler** [comment]
Please feel free :)

**Comment by karuppiah7890** [comment]
I have updated the PR #7062 ! :)

**Comment by karuppiah7890** [comment]
I have closed the PR #7062 . If anyone is interested to work on this issue, please go ahead 😄 Before getting started, do discuss with the maintainers on what's needed to fix this issue and check the comments on this issue and the PR too :)

**Comment by bacongobbler** [comment]
given feedback that was provided in #7062, I'm inclined to close this issue out as a wontfix. The use cases laid out in this ticket do not align here, as the platform, compiler and Go version all relate to the API server, not to the worker nodes. This was mentioned earlier at https://github.com/helm/helm/issues/7004#issuecomment-613050271 as well.

I'll let others chime in here to see if there's any use for these fields, but it's my belief that these are reporting information that chart authors are consuming under false pretenses that these reflect the node where the objecst are being scheduled onto.

@mmouly with regards to your use case mentioned in https://github.com/helm/helm/issues/7004#issuecomment-616390380, I'm not sure that is the correct way to determine which images you should use. Again, the capability fields are reporting compiler information on the API server. This may be better served using [multi-arch images](https://www.docker.com/blog/multi-arch-images/) to ensure the correct architecture is used based on the worker node selected.

**Comment by bacongobbler** [comment]
In any case, I'll remove this from the milestone.
